### PR TITLE
Fix stringFirst/stringLast rollup during ingestion

### LINKED ISSUE #7243: stringFirst/stringLast crashes at aggregation time
### Affected Version

0.13.0-incubating (and maybe older versions)

### Description

See discussion on #5789 for background.

Set up a fresh install of Druid 0.13-incubating following the [quickstart](http://druid.io/docs/latest/tutorials/index.html).

Write this index spec:

```json
{
  "type" : "index",
  "spec" : {
    "dataSchema" : {
      "dataSource" : "wikipedia",
      "parser" : {
        "type" : "string",
        "parseSpec" : {
          "format" : "json",
          "dimensionsSpec" : {},
          "timestampSpec": {
            "column": "time",
            "format": "iso"
          }
        }
      },
    "metricsSpec": [{
      "name": "channel",
      "fieldName": "channel",
      "type": "stringFirst",
      "maxStringBytes": 100
    }],
      "granularitySpec" : {
        "type" : "uniform",
        "segmentGranularity" : "day",
        "queryGranularity" : "hour",
        "intervals" : ["2015-09-12/2015-09-13"],
        "rollup" : true
      }
    },
    "ioConfig" : {
      "type" : "index",
      "firehose" : {
        "type" : "local",
        "baseDir" : "quickstart/tutorial/",
        "filter" : "wikiticker-2015-09-12-sampled.json.gz"
      },
      "appendToExisting" : false
    },
    "tuningConfig" : {
      "type" : "index",
      "targetPartitionSize" : 5000000,
      "maxRowsInMemory" : 1000,
      "forceExtendableShardSpecs" : true
    }
  }
}
```

and post it:
```
$ bin/post-index-task --file stringfirst-index.json
Beginning indexing data for wikipedia
Waiting up to 119s for indexing service [http://localhost:8090/] to become available. [Got: <urlopen error [Errno 61] Connection refused> ]
Task started: index_wikipedia_2019-03-12T19:07:26.507Z
Task log:     http://localhost:8090/druid/indexer/v1/task/index_wikipedia_2019-03-12T19:07:26.507Z/log
Task status:  http://localhost:8090/druid/indexer/v1/task/index_wikipedia_2019-03-12T19:07:26.507Z/status
Task index_wikipedia_2019-03-12T19:07:26.507Z still running...
Task index_wikipedia_2019-03-12T19:07:26.507Z still running...
Task index_wikipedia_2019-03-12T19:07:26.507Z still running...
Task index_wikipedia_2019-03-12T19:07:26.507Z still running...
Task finished with status: FAILED
```

The log reads:

```
2019-03-12T19:07:45,892 WARN [appenderator_merge_0] org.apache.druid.segment.realtime.appenderator.AppenderatorImpl - Failed to push merged index for segment[wikipedia_2015-09-12T00:00:00.000Z_2015-09-13T00:00:00.000Z_2019-03-12T19:07:26.636Z].
java.lang.ClassCastException: org.apache.druid.query.aggregation.SerializablePairLongString cannot be cast to java.lang.String
	at org.apache.druid.query.aggregation.first.StringFirstAggregateCombiner.reset(StringFirstAggregateCombiner.java:35) ~[druid-processing-0.13.0-incubating.jar:0.13.0-incubating]
	at org.apache.druid.segment.RowCombiningTimeAndDimsIterator.resetCombinedMetrics(RowCombiningTimeAndDimsIterator.java:249) ~[druid-processing-0.13.0-incubating.jar:0.13.0-incubating]
	at org.apache.druid.segment.RowCombiningTimeAndDimsIterator.combineToCurrentTimeAndDims(RowCombiningTimeAndDimsIterator.java:229) ~[druid-processing-0.13.0-incubating.jar:0.13.0-incubating]
	at org.apache.druid.segment.RowCombiningTimeAndDimsIterator.moveToNext(RowCombiningTimeAndDimsIterator.java:191) ~[druid-processing-0.13.0-incubating.jar:0.13.0-incubating]
	at org.apache.druid.segment.IndexMergerV9.mergeIndexesAndWriteColumns(IndexMergerV9.java:492) ~[druid-processing-0.13.0-incubating.jar:0.13.0-incubating]
	at org.apache.druid.segment.IndexMergerV9.makeIndexFiles(IndexMergerV9.java:191) ~[druid-processing-0.13.0-incubating.jar:0.13.0-incubating]
	at org.apache.druid.segment.IndexMergerV9.merge(IndexMergerV9.java:914) ~[druid-processing-0.13.0-incubating.jar:0.13.0-incubating]
	at org.apache.druid.segment.IndexMergerV9.mergeQueryableIndex(IndexMergerV9.java:832) ~[druid-processing-0.13.0-incubating.jar:0.13.0-incubating]
	at org.apache.druid.segment.IndexMergerV9.mergeQueryableIndex(IndexMergerV9.java:810) ~[druid-processing-0.13.0-incubating.jar:0.13.0-incubating]
	at org.apache.druid.segment.realtime.appenderator.AppenderatorImpl.mergeAndPush(AppenderatorImpl.java:719) ~[druid-server-0.13.0-incubating.jar:0.13.0-incubating]
	at org.apache.druid.segment.realtime.appenderator.AppenderatorImpl.lambda$push$1(AppenderatorImpl.java:623) ~[druid-server-0.13.0-incubating.jar:0.13.0-incubating]
	at com.google.common.util.concurrent.Futures$1.apply(Futures.java:713) [guava-16.0.1.jar:?]
	at com.google.common.util.concurrent.Futures$ChainingListenableFuture.run(Futures.java:861) [guava-16.0.1.jar:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) [?:1.8.0_162]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) [?:1.8.0_162]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_162]
```

@gianm suggested in #5789 that the AggregateCombiner code was just not running at all and that it should always be acting on SerializablePairLongString values rather than Strings. I am not enough of an expert on aggregation to know if that's correct.

Note that I needed to ensure that the task spec specified rollup: true, a large enough queryGranularity so that multiple rows do get combined, and a small enough maxRowsInMemory so that multiple spill files get combined.

### PR DESCRIPTION
**Pull request by joykent99**
Fixes #7243 .

### Description

This PR fixes the issues with stringFirst/stringLast rollup during ingestion, based on the suggestions described in #7243. A test is added to trigger the rollup code path for stringFirst/stringLast.

This PR also enables earliest/latest aggregator to operate on these columns during query time.

#### Added a forbidden-apis entry ...

<hr>

This PR has:
- [ ] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [ ] added documentation for new or modified features or behaviors.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/licenses.yaml)
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [ ] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [ ] been tested in a test Druid cluster.

<hr>

### CODE REVIEW

**Comment by joykent99** [comment]
This is the first time I submitted a patch to Druid. Any comments and suggestions are appreciated!

**Comment by suneet-s** [comment]
@joykent99 Thanks for the contribution! I haven't looked through the approach yet, but I notice there aren't any integration tests added for this change. Can you add one for the bug reported in #7243. This way we can be sure not to break this feature going forward.

I think you can do this by modifying one of the tests in `ITIndexerTest`. Consider adding a stringFirst and stringLast aggregation to `/indexer/wikipedia_index_task.json` and then modifying one of the queries in `/indexer/wikipedia_index_queries.json` to test that the aggregation is correct.

**Comment by joykent99** [comment]
Thanks @suneet-s for the tip on adding the integration tests. I will work on it and update the PR.

**Comment by joykent99** [comment]
@suneet-s, I have added some integration tests in this PR. It looks like the CI build failed due to license errors. Do you have any suggestions to how to resolve these errors? Thanks.

**Comment by suneet-s** [comment]
> @suneet-s, I have added some integration tests in this PR. It looks like the CI build failed due to license errors. Do you have any suggestions to how to resolve these errors? Thanks.

I re-triggered the job. I've seen it be flaky before, not sure why. Looks like everything is passing now 🎉 

I'll try to get some time to review this in the next couple days

File: integration-tests/src/test/resources/indexer/wikipedia_merge_index_task.json
Comment (suneet-s) [inline integration-tests/src/test/resources/indexer/wikipedia_merge_index_task.json:35]: Can we add these to `wikipedia_index_task.json` instead. This way we don't need to run another integration test which can be quite slow

You will probably want to do something very similar to #9277

File: processing/src/main/java/org/apache/druid/query/aggregation/first/StringFirstAggregatorFactory.java
Comment (suneet-s) [inline processing/src/main/java/org/apache/druid/query/aggregation/first/StringFirstAggregatorFactory.java:191]: wow! Is there a test for this somewhere?

File: sql/src/main/java/org/apache/druid/sql/calcite/aggregation/builtin/EarliestLatestAnySqlAggregator.java
Comment (suneet-s) [inline sql/src/main/java/org/apache/druid/sql/calcite/aggregation/builtin/EarliestLatestAnySqlAggregator.java:260]: We should reuse this instance across all calls to `toDruidAggregation`

File: sql/src/main/java/org/apache/druid/sql/calcite/aggregation/builtin/EarliestLatestAnySqlAggregator.java
Comment (suneet-s) [inline sql/src/main/java/org/apache/druid/sql/calcite/aggregation/builtin/EarliestLatestAnySqlAggregator.java:267]: shouldn't this remain `STRING` ?

**Review by suneet-s — COMMENTED** [review]
I don't fully understand how Combiners work, so I'll need to spend more time thinking about how this might impact other things. But I commented on stuff I have a little bit of a better understanding of.

**Comment by joykent99** [comment]
> I don't fully understand how Combiners work, so I'll need to spend more time thinking about how this might impact other things. But I commented on stuff I have a little bit of a better understanding of.

Thanks for taking a look, @suneet-s!

File: integration-tests/src/test/resources/indexer/wikipedia_merge_index_task.json
Comment (joykent99) [reply integration-tests/src/test/resources/indexer/wikipedia_merge_index_task.json:35]: The issue described in #7243 requires a trigger of merging of two index files, which will in turn cause a rollup of two rows. I wasn't able to trigger that in `wikipedia_index_task.json`. This is why I added a new task to capture this case.

File: processing/src/main/java/org/apache/druid/query/aggregation/first/StringFirstAggregatorFactory.java
Comment (joykent99) [reply processing/src/main/java/org/apache/druid/query/aggregation/first/StringFirstAggregatorFactory.java:191]: Yes. The existing test case was testing the wrong behavior. The fix to the test is also in this PR.

File: sql/src/main/java/org/apache/druid/sql/calcite/aggregation/builtin/EarliestLatestAnySqlAggregator.java
Comment (joykent99) [reply sql/src/main/java/org/apache/druid/sql/calcite/aggregation/builtin/EarliestLatestAnySqlAggregator.java:260]: Sounds good.

File: sql/src/main/java/org/apache/druid/sql/calcite/aggregation/builtin/EarliestLatestAnySqlAggregator.java
Comment (joykent99) [reply sql/src/main/java/org/apache/druid/sql/calcite/aggregation/builtin/EarliestLatestAnySqlAggregator.java:267]: When we specify "stringFirst/stringLast" in ingestion MetricsSpec, the type for the columns stored in segments is `OTHER`. With `STRING` type here, the SQL will report an error when trying to call `earliest/latest` on those columns.

Comment (joykent99) [reply sql/src/main/java/org/apache/druid/sql/calcite/aggregation/builtin/EarliestLatestAnySqlAggregator.java:267]: If you have any suggestions on where is a good place to add a test for this, that would be great!

File: integration-tests/src/test/resources/indexer/wikipedia_merge_index_task.json
Comment (joykent99) [reply integration-tests/src/test/resources/indexer/wikipedia_merge_index_task.json:35]: I also added these in `wikipedia_index_task.json`, which led me to discover the issue with `StringFirstAggregatorFactory.combine()` below.

File: processing/src/test/java/org/apache/druid/query/aggregation/first/StringFirstAggregationTest.java
Comment (joykent99) [inline processing/src/test/java/org/apache/druid/query/aggregation/first/StringFirstAggregationTest.java:120]: This is the fix to the existing test for `StringFirstAggregatorFactory.combine()` fixes.

**Comment by joykent99** [comment]
@suneet-s Please let me know if you have more comments on this PR. It looks like @gianm also found the issue with StringFirstAggregatorFactory.combine() function, which is what caused the above conflicts. Should I pull in the new changes in the master to my branch, and force push it? Thanks.

**Comment by suneet-s** [comment]
@joykent99 the way I resolve merge conflicts is like this

```
git fetch upstream
git merge upstream/master
# Resolve merge conflicts manually
# ...
git push origin <branch>
```

This way you don't need to force push, so the comment history is still referenceable in the PR. I think the change looks good to me, just going to do a once over today and post final comments.

Thanks again for fixing this!

File: sql/src/main/java/org/apache/druid/sql/calcite/aggregation/builtin/EarliestLatestAnySqlAggregator.java
Comment (suneet-s) [reply sql/src/main/java/org/apache/druid/sql/calcite/aggregation/builtin/EarliestLatestAnySqlAggregator.java:268]: Usually I sql add tests to `CalciteQueryTest` You can see `testEarliestAggregators` as an example. You could write a test in here to validate the SQL translation works, but I think you need to update `CalciteTests#INDEX_SCHEMA_DIFFERENT_DIM3_M1_TYPES` to add an earliest / latest aggregator. Then you should be able to test a SQL test in `CalciteQueryTest` - hope this helps

**Review by suneet-s — COMMENTED** [review]
LGTM after resolving merge conflicts.

**Comment by joykent99** [comment]
@suneet-s I have merged in the changes in master.

**Review by suneet-s — APPROVED** [review]
LGTM - thanks @joykent99 ! 🎉

**Comment by bavaria95** [comment]
@joykent99 does the line `First and Last aggregator cannot be used in ingestion spec, and should only be specified as part of queries.` in docs still correct after this MR?

**Comment by suneet-s** [comment]
> @joykent99 @suneet-s
> 
> does the line `First and Last aggregator cannot be used in ingestion spec, and should only be specified as part of queries.` in docs still correct after this MR?

@bavaria95 Can you point me to a link of the doc you mentioned please. See #7599 Long / Double / Float first/last aggs still can not be used at ingestion time

**Comment by bavaria95** [comment]
@suneet-s https://druid.apache.org/docs/latest/querying/aggregations.html#first--last-aggregator the first line here

**Comment by bavaria95** [comment]
@joykent99 but firstString/lastString should be working fine?

Could you please explain how to use it inside the query? Let's say we added a metric to be `lastString` aggregator during ingestion time. And this metric has `OTHER` type:

image


How can it be queried, something like say `SELECT * WHERE test='JarBot'`?

**Comment by joykent99** [comment]
> @joykent99 but firstString/lastString should be working fine?
> 
> Could you please explain how to use it inside the query? Let's say we added a metric to be `lastString` aggregator during ingestion time. And this metric has `OTHER` type:
> 
> How can it be queried, something like say `SELECT * WHERE test='JarBot'`?

Since the data for `lastString` aggregator stored in segments are in `SerializablePair`, to get the final result, we issue the query like this:

```
select <your_grouping_column>, latest(test, 1024)
from wikipedia-test
group by <your_grouping_column>
```

**Comment by bavaria95** [comment]
Thanks, I will try so!

**Pull request merged**
